air/src/proof/security.rs

```diff
@@ -0,0 +1,728 @@
+            + extension_field_bits;
+    epsilons_bits_neg.push(epsilon_2_bits_neg);
+
+    // Computes FRI commit-phase (i.e., pre-query) soundness error
+    let epsilon_3_bits_neg = extension_field_bits - log2(lde_domain_size);
+    epsilons_bits_neg.push(epsilon_3_bits_neg);
+
+    // epsilon_i for i in [3..(k-1)], where k is number of rounds
+    let folding_factor = options.to_fri_options().folding_factor() as f64;
+    let num_fri_layers = options.to_fri_options().num_fri_layers(lde_domain_size as usize);
+    let epsilon_i_min_bits_neg = (0..num_fri_layers)
+        .map(|_| extension_field_bits - log2((folding_factor - 1.0) * (lde_domain_size + 1.0)))
+        .fold(f64::INFINITY, |a, b| a.min(b));
+    epsilons_bits_neg.push(epsilon_i_min_bits_neg);
+
+    // Compute FRI query-phase soundness error
+    let epsilon_k_bits_neg = options.grinding_factor() as f64 - log2(powf(alpha, num_fri_queries));
+    epsilons_bits_neg.push(epsilon_k_bits_neg);
+
+    // return the round-by-round (RbR) soundness error
+    epsilons_bits_neg.into_iter().fold(f64::INFINITY, |a, b| a.min(b)) as u64
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Computes the largest proximity parameter m such that eta is greater than 0 in the proof of
+/// Theorem 1 in https://eprint.iacr.org/2021/582. See Theorem 2 in https://eprint.iacr.org/2024/1553
+/// and its proof for more on this point.
+///
+/// The bound on m in Theorem 2 in https://eprint.iacr.org/2024/1553 is sufficient but we can use
+/// the following to compute a better bound.
+fn compute_upper_m(h: usize) -> f64 {
+    let h = h as f64;
+    let ratio = (h + 2.0) / h;
+    let m_max = ceil(1.0 / (2.0 * (sqrt(ratio) - 1.0)));
+    assert!(m_max >= h / 2.0, "the bound in the theorem should be tighter");
+
+    // We cap the range to 1000 as the optimal m value will be in the lower range of [m_min, m_max]
+    // since increasing m too much will lead to a deterioration in the FRI commit soundness making
+    // any benefit gained in the FRI query soundess mute.
+    cmp::min(m_max as u64, MAX_PROXIMITY_PARAMETER) as f64
+}
+
+#[cfg(feature = "std")]
+pub fn log2(value: f64) -> f64 {
+    value.log2()
+}
+
+#[cfg(not(feature = "std"))]
+pub fn log2(value: f64) -> f64 {
+    libm::log2(value)
+}
+
+#[cfg(feature = "std")]
+pub fn sqrt(value: f64) -> f64 {
+    value.sqrt()
+}
+
+#[cfg(not(feature = "std"))]
+pub fn sqrt(value: f64) -> f64 {
+    libm::sqrt(value)
+}
+
+#[cfg(feature = "std")]
+pub fn powf(value: f64, exp: f64) -> f64 {
+    value.powf(exp)
+}
+
+#[cfg(not(feature = "std"))]
+pub fn powf(value: f64, exp: f64) -> f64 {
+    libm::pow(value, exp)
+}
+
+#[cfg(feature = "std")]
+pub fn ceil(value: f64) -> f64 {
+    value.ceil()
+}
+
+#[cfg(not(feature = "std"))]
+pub fn ceil(value: f64) -> f64 {
+    libm::ceil(value)
+}
+
+#[cfg(test)]
+mod prove_security_tests {
```

Comment:
nit: let's add a section separator above this line with title `TESTS`.

Also, we could probably rename the module to just `tests`.

Reply:
changed